### PR TITLE
gh-95913: Copyedit, link & format Typing Features section in 3.11 What's New

### PR DESCRIPTION
**Pull request by CAM-Gerlach**
Part of #95913

Fixes Sphinx and textual issues in the What's New in Python 3.11 "New Features Related to Type Hints" section, improves the clarity, quality and non-duplication of the text, and links to referenced documentation sections.

Non-trivial highlights:
* Split PEP 646 description into two paragraphs for easier readability
* Clarify the syntax and semantics of the description of the interaction of `total` and required
* Link the PEP 484 section implicitly referred by PEP 673 directly, as its quite non-trivial to find
* Use a more Pythonically (and perhaps culturally) first/last name in PEP 681 example
* Clarify PEP 563 status description, particularly that this is not the first deferral, the more common form as well as a link to the `__future__` statement, and make the link text more descriptive

General changes:
* Add Sphinx cross-references to additional items
* Clarify the text in various spots
* Fix/improve a number of other minor grammar, punctuation and phrasing issues
* Consistent section heading capitalization, ref target labels and spacing

* Issue: gh-95913

### CODE REVIEW

File: Doc/whatsnew/3.11.rst [hunk 1]
```diff
@@ -203,25 +203,31 @@ Irit Katriel in :issue:`45607`.)
...
 with an *arbitrary* number of types. In other words,
 a :data:`~typing.TypeVarTuple` is a *variadic* type variable,
-enabling *variadic* generics. This enables a wide variety of use cases.
+enabling *variadic* generics.
+
+This enables a wide variety of use cases.
 In particular, it allows the type of array-like structures
 in numerical computing libraries such as NumPy and TensorFlow to be
-parameterised with the array *shape*. Static type checkers will now
+parameterised with the array *shape*, and static type checkers will now
```
Comment (JelleZijlstra) [inline Doc/whatsnew/3.11.rst]: I prefer the previous version. Overly long sentences are less readable.

**Review by JelleZijlstra — COMMENTED** [review]
Thanks, mostly looks good!

File: Doc/whatsnew/3.11.rst [hunk 2]
```diff
@@ -230,26 +236,30 @@ See :pep:`646` for more details.
...
    class Movie(TypedDict):
       title: str
       year: NotRequired[int]
 
-   m1: Movie = {"title": "Black Panther", "year": 2018}  # ok
-   m2: Movie = {"title": "Star Wars"}  # ok (year is not required)
-   m3: Movie = {"year": 2022}  # error (missing required field title)
+   m1: Movie = {"title": "Black Panther", "year": 2018}  # OK
+   m2: Movie = {"title": "Star Wars"}  # OK (year is not required)
+   m3: Movie = {"year": 2022}  # ERROR (missing required field title)
```
Comment (AlexWaygood) [inline Doc/whatsnew/3.11.rst:262]: What's the motivation for the ALLCAPS change here? It looks like shouting to me.

("OK" is fine but I'm not a fan of "ERROR".)

File: Doc/whatsnew/3.11.rst [hunk 3]
```diff
@@ -262,15 +272,20 @@ See :pep:`655` for more details.
...
 way to annotate methods that return an instance of their class. This
-behaves the same as the :data:`~typing.TypeVar`-based approach specified
-in :pep:`484` but is more concise and easier to follow.
+behaves the same as the :data:`~typing.TypeVar`-based approach
+:pep:`specified in PEP 484<484#annotating-instance-and-class-methods>`,
+but is more concise and easier to follow.
 
-Common use cases include alternative constructors provided as classmethods
+Common use cases include alternative constructors provided as
+:func:`classmethod <classmethod>`\s,
```
Comment (AlexWaygood) [inline Doc/whatsnew/3.11.rst:288]:
```suggestion
:func:`classmethods <classmethod>`,
```

File: Doc/whatsnew/3.11.rst [hunk 4]
```diff
@@ -329,14 +347,17 @@ See :pep:`675` for more details.
...
 PEP 681: Data Class Transforms
 ------------------------------
 
 :data:`~typing.dataclass_transform` may be used to
 decorate a class, metaclass, or a function that is itself a decorator.
 The presence of ``@dataclass_transform()`` tells a static type checker that the
-decorated object performs runtime "magic" that
-transforms a class, giving it :func:`dataclasses.dataclass`-like behaviors.
+decorated object performs runtime "magic" that transforms a class,
+giving it :func:`dataclass <dataclasses.dataclass>`-like behaviors.
```
Comment (AlexWaygood) [inline Doc/whatsnew/3.11.rst:360]:
```suggestion
giving it :func:`~dataclasses.dataclass`-like behaviors.
```

File: Doc/whatsnew/3.11.rst [hunk 5]
```diff
@@ -348,26 +369,31 @@ For example::
...
+
+PEP 563 may not be the future
 -----------------------------
 
-* :pep:`563` Postponed Evaluation of Annotations, ``__future__.annotations``
-  that was planned for this release has been indefinitely postponed.
-  See `this message <https://mail.python.org/archives/list/python-dev@python.org/message/VIZEBX5EYMSYIJNDBF6DMUMZOCWHARSO/>`_ for more information.
+:pep:`563` Postponed Evaluation of Annotations
+(the ``from __future__ import annotations`` :ref:`future statement <future>`)
+that was originally planned for release in Python 3.10 has been deferred again.
```
Comment (AlexWaygood) [inline Doc/whatsnew/3.11.rst]: I prefer the current phrasing of "indefinitely postponed" rather than "has been deferred again"; it feels like it describes the situation better.

"Has been deferred again" implies that the expectation is still that it will become the default in a future Python version. That isn't true anymore; we don't know what the default will be. It probably won't be the status quo, but it might be PEP 649, or something else entirely.

Comment (JelleZijlstra) [reply Doc/whatsnew/3.11.rst]: Yes, "deferred again" suggests that PEP 563 will become the default in 3.12, which isn't looking likely to me.

File: Doc/whatsnew/3.11.rst (hunk 1 above)
Comment (CAM-Gerlach) [reply Doc/whatsnew/3.11.rst]: Reverted.

File: Doc/whatsnew/3.11.rst (hunk 2 above)
Comment (CAM-Gerlach) [reply Doc/whatsnew/3.11.rst:262]: It is intended to draw emphasis to the key contrast these examples are included to illustrate—the first and second are valid, while the third is invalid and produces an error. Also, "OK" is of course the correct, canonical capitalization of "OK", while it is "Error" is ALLCAPSed given the particular importance and best practice of prominently marking bad code as such.

File: Doc/whatsnew/3.11.rst (hunk 3 above)
Comment (CAM-Gerlach) [reply Doc/whatsnew/3.11.rst:288]: I initially had this, but it results in the whole word `classmethods` being formatted in the literal font, which is not correct—`classmethods` is not the literal name of the linked function as the literal formatting incorrectly implies. What would be better is to just link to it as a regular prose word, like a `:ref:` (which is what is more appropriate here anyway), but that would require adding a dedicated inbound ref target label just for this, which I'm not sure is worth it. Thoughts?

File: Doc/whatsnew/3.11.rst (hunk 4 above)
Comment (CAM-Gerlach) [reply Doc/whatsnew/3.11.rst:360]: Likewise I initially had this, but without a custom link text, it appends a `()` to the name, which is both syntactically disruptive and not semantically correct here, as it is referring to the objects decorated _by_ the dataclass decorator, not the dataclass function itself.

File: Doc/whatsnew/3.11.rst (hunk 5 above)
Comment (CAM-Gerlach) [reply Doc/whatsnew/3.11.rst]: I restored the original "indefinitely" wording, only retaining "deferred" instead of "postponed". The reason I changed "postponed" to begin with was due to word repetition and the significant potential for confusion due to the title and subject of the PEP in question being "postponed" evaluation of annotations (it even confused me for a moment).

File: Doc/whatsnew/3.11.rst (hunk 4 above)
Comment (AlexWaygood) [reply Doc/whatsnew/3.11.rst:360]: Ah, fair enough!

File: Doc/whatsnew/3.11.rst (hunk 3 above)
Comment (AlexWaygood) [reply Doc/whatsnew/3.11.rst:288]: Your rationale makes sense; let's stick with your version in that case

File: Doc/whatsnew/3.11.rst (hunk 2 above)
Comment (AlexWaygood) [reply Doc/whatsnew/3.11.rst:262]: "ERROR" still seems ugly and unnecessary to me; I don't really see the need for a change here.

File: Doc/whatsnew/3.11.rst [hunk 6]
```diff
@@ -348,26 +369,32 @@ For example::
...
+PEP 563 may not be the future
 -----------------------------
 
-* :pep:`563` Postponed Evaluation of Annotations, ``__future__.annotations``
-  that was planned for this release has been indefinitely postponed.
-  See `this message <https://mail.python.org/archives/list/python-dev@python.org/message/VIZEBX5EYMSYIJNDBF6DMUMZOCWHARSO/>`_ for more information.
+:pep:`563` Postponed Evaluation of Annotations
+(the ``from __future__ import annotations`` :ref:`future statement <future>`)
+that was originally planned for release in Python 3.10
+has been deferred indefinitely.
```
Comment (AlexWaygood) [inline Doc/whatsnew/3.11.rst]: "Deferred" still sounds to me like there's a strong presumption it will happen eventually.

```suggestion
has been put on hold indefinitely.
```

File: Doc/whatsnew/3.11.rst (hunk 2 above)
Comment (AlexWaygood) [reply Doc/whatsnew/3.11.rst:262]: It also goes against the prevailing style in the `typing` docs — there are several `# error` comments,  but none of them are ALLCAPS: https://docs.python.org/3/library/typing.html

File: Doc/whatsnew/3.11.rst [hunk 7]
```diff
@@ -262,15 +272,20 @@ See :pep:`655` for more details.
...
+.. _whatsnew311-pep673:
+
 PEP 673: ``Self`` type
 ----------------------
 
 The new :data:`~typing.Self` annotation provides a simple and intuitive
 way to annotate methods that return an instance of their class. This
-behaves the same as the :data:`~typing.TypeVar`-based approach specified
-in :pep:`484` but is more concise and easier to follow.
+behaves the same as the :data:`~typing.TypeVar`-based approach
```
Comment (ezio-melotti) [inline Doc/whatsnew/3.11.rst]: Should this be
```suggestion
behaves the same as the :class:`~typing.TypeVar`-based approach
```
as well?

File: Doc/whatsnew/3.11.rst [hunk 8]
```diff
@@ -262,15 +272,20 @@ See :pep:`655` for more details.
...
+:pep:`specified in PEP 484<484#annotating-instance-and-class-methods>`,
```
Comment (ezio-melotti) [inline Doc/whatsnew/3.11.rst]:
```suggestion
:pep:`specified in PEP 484 <484#annotating-instance-and-class-methods>`,
```

File: Doc/whatsnew/3.11.rst (hunk 4 above)
Comment (CAM-Gerlach) [reply Doc/whatsnew/3.11.rst:360]: In other words, "great minds"... :)

File: Doc/whatsnew/3.11.rst (hunk 2 above)
Comment (CAM-Gerlach) [reply Doc/whatsnew/3.11.rst:262]: Personally, I value clarity and explicitness over subjective aesthetic preference, given a stated rationale. Also, I'm not sure how relevant the conventions of the `typing` stdlib doc are here, given it is a separate document from this one, and those particular choices may or may not have a clear, thought-out rationale.

But what do others think? @ezio-melotti @JelleZijlstra ?

If an all-caps `ERROR` is simply too objectionable, we can at least give it proper case

```suggestion
   m3: Movie = {"year": 2022}  # Error (missing required field title)
```

File: Doc/whatsnew/3.11.rst (hunk 7 above)
Comment (CAM-Gerlach) [reply Doc/whatsnew/3.11.rst]: Indeed it should—I missed a spot here. Thanks!

File: Doc/whatsnew/3.11.rst (hunk 8 above)
Comment (CAM-Gerlach) [reply Doc/whatsnew/3.11.rst]: Oops, don't know how I did that...rather embarrassing. Thanks!

File: Doc/whatsnew/3.11.rst (hunk 6 above)
Comment (CAM-Gerlach) [reply Doc/whatsnew/3.11.rst]: To me "deferred indefinitely" didn't read as any stronger of a presumption than "postponed indefinitely", but if "put on hold indefinitely" reads closer to the intend to you, we can just go with that :+1:

**Review by AlexWaygood — APPROVED** [review]
Thanks! I still don't really think there's a good rationale for changing `# error` to `# ERROR`, but that's a minor nit :)

File: Doc/whatsnew/3.11.rst (hunk 2 above)
Comment (ezio-melotti) [reply Doc/whatsnew/3.11.rst:262]: I don't have a strong opinion either way.

For inline comments like these I tend to use all lowercase and no full stops at the end, so even `ok` and `error` would be fine with me.
I don't see a strong reason to emphasize the `ERROR` since each example has a matching comment -- it's not a warning that needs to be made more noticeable otherwise people might miss it.

OTOH I would rephrase it to
```suggestion
   m3: Movie = {"year": 2022}  # ERROR (title is required, but missing)
```
which follows the previous example more closely and is IMHO more readable (with whatever case you settle on for `ERROR`).

Comment (CAM-Gerlach) [reply Doc/whatsnew/3.11.rst:262]: @AlexWaygood what about if we go with @ezio-melotti 's suggestion and not ALLCAPSing Error, like this?

```suggestion
   m3: Movie = {"year": 2022}  # Error (title is required, but missing)
```

Comment (AlexWaygood) [reply Doc/whatsnew/3.11.rst:262]: > @AlexWaygood what about if we go with @ezio-melotti 's suggestion and not ALLCAPSing Error, like this?

Sounds good to me.

Comment (CAM-Gerlach) [reply Doc/whatsnew/3.11.rst:262]: > For inline comments like these I tend to use all lowercase and no full stops at the end, so even ok and error would be fine with me.

Same, I don't add a period at the end either for inline or single-line comments...but I notice you _do_ use period for titles (and commit summary lines), which slightly confuses me :)  (To note, I do start them with a capital letter, which follows [the examples in PEP 8](https://peps.python.org/pep-0008/#comments) and avoids looking (IMO) sloppy—though that latter point is personal aesthetic preference)

Comment (encukou) [reply Doc/whatsnew/3.11.rst:262]: FWIW, in this general document, I'd appreciate a subtle indication that this is a *typing* error -- not a SyntaxError or runtime exception. Perhaps `Wrong`, in whatever case you prefer.

**Comment by encukou** [comment]
Merging, the minor nit can be resolved later.

File: Doc/whatsnew/3.11.rst (hunk 2 above)
Comment (CAM-Gerlach) [reply Doc/whatsnew/3.11.rst:262]: That's a great point—to be more specific, we could say something like `Type check error` or `Incompatible type`
